client.go

```diff
@@ -282,50 +303,73 @@ func (client *HTTPClient) PatchForBody(url string, requestBody interface{}, resp
 	if err != nil {
 		return err
 	}
-	defer resp.Body.Close()
+	defer resp.Raw.Body.Close()
 
-	dec := json.NewDecoder(resp.Body)
-	jsonErr := dec.Decode(responseBody)
+	decodeErr := resp.decodeBody(responseBody)
 
-	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusCreated {
-		return HTTPClientError{Code: resp.StatusCode, Status: resp.Status}
+	if resp.Raw.StatusCode != http.StatusOK && resp.Raw.StatusCode != http.StatusCreated {
+		return HTTPClientError{Code: resp.Raw.StatusCode, Status: resp.Raw.Status}
 	}
 
-	if jsonErr != nil {
-		return errors.Wrap(jsonErr, "decoding response body")
+	if decodeErr != nil {
+		return errors.Wrap(decodeErr, "decoding response body")
 	}
 
-	// Read all additional bytes from the body
-	ioutil.ReadAll(resp.Body)
-
 	return nil
 }
 
-func (client *HTTPClient) perform(method string, url string, requestBody interface{}, params ...RequestParam) (*http.Response, error) {
-	data, err := json.Marshal(requestBody)
+func (client *HTTPClient) perform(method string, url string, requestBody interface{}, params ...RequestParam) (response, error) {
+	rawReq, err := http.NewRequest(method, url, nil)
 	if err != nil {
-		return nil, errors.Wrap(err, "marshalling entity")
+		return response{}, errors.Wrapf(err, "invalid url %v", url)
 	}
 
-	req, err := http.NewRequest(method, url, bytes.NewBuffer(data))
+	req := defaultRequest(rawReq)
+
+	data, err := req.Encode(requestBody)
 	if err != nil {
-		return nil, errors.Wrapf(err, "invalid url %v", url)
+		return response{}, errors.Wrap(err, "marshalling entity")
 	}
 
+	params = append(params, ContentTypeJSON())
 	client.applyParams(req, params)
 
-	req.Header.Add("Content-Type", "application/json")
+	req.Raw.Body = ioutil.NopCloser(bytes.NewBuffer(data))
 
-	resp, err := client.wrapped.Do(req)
+	rawResp, err := client.wrapped.Do(req.Raw)
 	if err != nil {
-		return nil, errors.Wrapf(err, "request failed %v", url)
+		return response{}, errors.Wrapf(err, "request failed %v", url)
 	}
 
-	return resp, nil
+	return response{
+		Raw:    rawResp,
+		Decode: req.Decode,
+	}, nil
 }
 
-func (client *HTTPClient) applyParams(req *http.Request, params []RequestParam) {
+func (client *HTTPClient) applyParams(req Request, params []RequestParam) {
```

Comment:
Muss das nicht auch ein Pointer sein?
```suggestion
func (client *HTTPClient) applyParams(req *Request, params []RequestParam) {
```